### ISSUE TITLE
Task: Error propagation

### ISSUE THREAD

**Issue by hbushouse** [STScI-JWST/jwst#596]
The JCCWG is close to agreement on a baseline scheme for tracking and propagating various error components through pipeline processing. Their work on the subject is being tracked at https://confluence.stsci.edu/display/JWSTCC/Optimal+Error+Initialization and discussion from a recent meeting on the subject is available at https://confluence.stsci.edu/display/JWSTCC/2017-01-10+Meeting+notes

Implementation of some form of error propagation is required by DMS Level 4 requirement 464.

**Comment by hbushouse** [comment]
Documenting some email discussion, based on a conversation between Karl and Phil in July. He wants to support the following approach:

Store variances rather than error estimates.

At the very least keep read noise and photon (Poisson) noise.

The ramp_fit step should be modified to do the following:

(1) Fit the slope using ordinary least squares with optimal weighting.  [Note that this may actually be done multiple times for each pixel, once for each segment between jumps.  This is a significant complication.]  The weighting (i.e. the weight as a function of group number) depends only on the signal-to-noise ratio of the data.  The weighting function has an expression raised to a power, and that power is obtained from a table using the SNR.  Initially, use all available variance estimates (currently only read noise and photon noise) to compute the signal-to-noise ratio.  Do the fit; save the value of the slope. Also compute the formal variance (i.e. based on the input variances, not based on the scatter of the data about the fitted line), but this is only used for weighting the segments when averaging to get the slope for each integration and for weighting integrations to get the overall slope for the exposure.

[So far, I think this is what is currently done, but using only the read noise and photon noise.]

Do step (2) for each of the available variance estimates.

(2) Re-compute the slope using OLS and optimal weighting as in step (1), but when computing the signal-to-noise ratio use only the current variance estimate.  Ignore the fitted slope but compute and save the formal variance.

The variances (so far) are additive.  The full variance of the slope should be the sum of all these individual variances from iterating step (2); this "full variance" is the variance that we got in step (1) and discarded.

A pipeline or step that combines or further processes slope images should compute the variance of the output, and that pipeline or step would (with the modification outlined above) have a choice about which variances in (or with) the _rate files to use when computing the output variance.

Complications:

The read noise is no problem.  Note that it's a 2-D array, dependent only on the detector.  The photon variance array might be 3-D:  integrations, rows, columns.  Other variance arrays might be 4-D.  Storing these arrays in the _ramp file or _rate file would be pretty unweildy.  (Consider TSO data.)

For the variance due to photon noise, we can't just take the counts (divided by gain) at the end of the integration, because that will likely include cosmic-ray jumps.  Also, the photon noise is needed segment-by-segment, when fitting a line to each segment.  The ramp_fit step gets the photon noise by subtracting the data at the first group in a segment from the last group in the segment.  It seems to me that we should leave that alone; trying to do that separately from ramp_fit would be a can of worms.  Note that this is different from saving the variances in the slope that were computed using just the photon noise; we need to do that.

Another difficulty with storing these arrays in a _ramp or _rate file is that those are rigid formats; the models would have to be updated to explicitly list every variance array that we want to save.  (Well, it looks as if we could use the "extra_fits" dictionary, but somehow that doesn't seem like a good idea.)

It would work (I tried this) to define a MultiVarianceModel, that can hold a list of VarianceModel arrays.  If the schema for the VarianceModel does not specify a dimension, such a Multi could be used to contain arrays of different shapes (e.g. read noise and photon noise), and we would get a FITS file with multiple image extensions.  The problem is that we would then need to associate this file of variances with the _ramp or _rate file.
